### DISCUSSION TITLE
DevOps i agent w linuxie. Problem z EF Core

### DISCUSSION THREAD

**Discussion by szympulka** [AzureCommunityPL/group#81] [Pomoc] [answered]
Hej,

Z reguły tworzyłem agenta na windowsie. Z czasem zmieniły się wymagania i stawiam wszystko na ubuntu. Odpalam coomand line w konkretnym working directory ze skryptem: 

dotnet ef migrations script --idempotent -o $(Build.ArtifactStagingDirectory)/db.sql

Agent wypluwa mi błąd jak na zdjęciu. Gdy robię to za pomocą ssh to wszystko śmiga. Gdzie może leżeć problem?

![image](https://user-images.githubusercontent.com/12598823/93972468-9bc99580-fd72-11ea-9b2b-551e73a77786.png)

**Comment by ptrstpp950** [comment]
Obstawiam 2 możliwe problemy:
1) SSH i Agent chodzą na innym użytkowniku
2) Zestaw "toolset" dotnet core jest nadpisywany przez agenta

Najprościej dodaj do skryptu: `dotnet tool install --global dotnet-ef`
A jak chcesz to zrobić porządnie to sprawdź czy jest dostępne dotnet ef a jak nie to zainstaluj

**Comment by szympulka** [comment, marked as answer]
Zainstalowałem go już globalnie, ale idąc Twoją radą, zrobiłem po najmniejszej linii oporu -> 
![image](https://user-images.githubusercontent.com/12598823/93985546-8742c880-fd85-11ea-9a7b-ec4a83448b98.png)

Agent powiedział, że jest zainstalowany a następnie się wywalił

**Reply by kamil-mrzyglod** [reply]
Dwa pytania:
* jest szansa abyś pokazał jak dokładnie tego shella odpalasz?
* to jest hosted czy self-hosted agent?

**Reply by szympulka** [reply]
Odpalam to jako "Command line" i jest to self-hosted agent

**Reply by ptrstpp950** [reply]
Ale czy możesz wkleić nam ten command line albo potwierdzić że wygląda on tak:
```
dotnet tool install --global dotnet-ef
dotnet ef migrations script --idempotent -o $(Build.ArtifactStagingDirectory)/db.sql
```
Jeżeli jest coś jeszcze, jest coś inaczej, itd. To prośba wklej.

**Reply by szympulka** [reply]
Dokładnie tak wygląda

**Reply by szympulka** [reply]
I tutaj odpalone na maszynie 
![image](https://user-images.githubusercontent.com/12598823/93994935-b6126c00-fd90-11ea-86ef-5b32e7c6238b.png)

**Reply by ptrstpp950** [reply]
Dołóż sobie do skryptu `ls` bo coś czuję że brakuje pliku SQL w katalogu a komunikat jest po prostu do chrzanu

**Reply by szympulka** [reply]
Jak widać, jestem w dobrym katalogu (widać nawet stary plik sql)
![image](https://user-images.githubusercontent.com/12598823/94000492-84e96a00-fd97-11ea-9f2a-4d694be7edf5.png)

**Reply by kamil-mrzyglod** [reply]
IMHO to jest problem z tym, że w PATH użytkownika z którego korzystasz w DevOps nie ma migratora do EF pomimo instalacji(a jak logujesz się po SSH to prawdopodobnie tam jednak ścieżka jest). Spróbuj:
```
dotnet tool install --global dotnet-ef
export PATH="$PATH:$HOME/.dotnet/tools/"
dotnet ef
```
Z drugiej strony jeśli to self-hosted to zamiast się tutaj w CMD bawić, może po prostu zainstaluj na agencie tego toola do EF i upewnij sie, że jest poprawnie wyeksportowany do PATH?

**Reply by szympulka** [reply]
No i teraz śmiga. Chociaż dodawałem PATH gdy logowałem się do maszyny przez ssh

**Reply by kamil-mrzyglod** [reply]
Ale to nie przypadkiem dla innego usera niż ten, na którego loguje się DevOps? :)

**Reply by szympulka** [reply]
@kamil-mrzyglod ale ciekawe jest to, że za każdym razem muszę dodwać export PATH="$PATH:$HOME/.dotnet/tools/"
 do skryptu. Czy on zmienia sobie zmienną środowiskową za każdym razem lub ją usuwa?

**Reply by szympulka** [reply]
Co do samego usera. Przy agencie na win faktycznie pytało mnie o użytkownika. Na linuxie sudo i w sumie tyle

**Reply by kamil-mrzyglod** [reply]
Może to jest jakiś bug z tymi toolsami - może spróbować to zgłosić na przyszłość. Ja osobiście nie używam EF więc ciężko mi się wypowiedzieć :)
